src/lib/es5.d.ts

```diff
@@ -1507,11 +1507,7 @@ interface DataViewConstructor {
 }
 declare const DataView: DataViewConstructor;
 
-/**
-  * A typed array of 8-bit integer values. The contents are initialized to 0. If the requested
-  * number of bytes could not be allocated an exception is raised.
-  */
-interface Int8Array {
+interface TypedArray<T> {

```

Comment:
What value does the generic `T` give you here? You could just use the `this` type wherever you use `T` below without making `TypedArray` generic.

Reply:
See https://github.com/Microsoft/TypeScript/issues/18425

Reply:
and slice and subarray don't necessarily return this if I'm not mistaken

Reply:
I think we have since fixed some of the underlying issues that made it slower.

Reply:
> and slice and subarray don't necessarily return this if I'm not mistaken

that i do not know how to address with this...

Reply:
`slice` uses [TypedArraySpeciesCreate](https://tc39.github.io/ecma262/#typedarray-species-create), per [22.2.3.24](https://tc39.github.io/ecma262/#sec-%typedarray%.prototype.slice), step 9.

Similarly, `subarray` also uses TypedArraySpeciesCreate, per [22.2.3.27](https://tc39.github.io/ecma262/#sec-%typedarray%.prototype.subarray), step 17.

TypedArraySpeciesCreate creates a new %TypedArray% instance using the current subclass constructor. As such, having both methods return `this` would be correct.

Reply:
While they don't exactly return the `this` reference, they do return a subclass instance of the same type as `this`, which probably is close enough since we can't do `this["constructor"]["prototype"]`.

Reply:
@rbuckton https://github.com/Microsoft/TypeScript/pull/18652#issuecomment-337051589

Reply:
That comment was specific to `[Symbol.species]`. I can see how that could apply to an attempt to define a shared `TypedArrayConstructor` interface, but not to the `TypedArray` instance type.

Alternatively, define `TypedArray` with `this` **except** for `slice`/`subarray` and have the specific instance types override `slice`/`subarray` with their specific return types.

Reply:
Yes, the thisArg params can be typed with `this`. Is there a disadvantage to using T for for the slice/subarray return types?

Reply:
We can't properly constrain the generic, so you could write a `TypedArray<HTMLElement>` which would not make sense.

Reply:
As nice as it would be to completely deduplicate `TypedArray`, I don't believe that is a completely viable option at this point. As we cannot properly constrain `T`, I would recommend we do not make `TypedArray` generic at this time.

Reply:
Having TypedArray generic allows for the following construct, which would type all the methods which use SpeciesConstructor correctly.
```
class MyInt8Array extends (Int8Array as TypedArray<MyInt8Array>) {
    [Symbol.species]: MyInt8Array
}
```

Reply:
What's the issue with `interface TypedArray<T extends Uint16Array | Uint8Array | ...> {` if you want to constrain TypedArray?